src/peer_transport/transport.ts

```diff
@@ -0,0 +1,154 @@
+export const TRANSPORT = '/p2p-webrtc-direct'
+export const PROTOCOL = '/webrtc-signaling/0.0.1'
+export const CODE = 276
+
+export interface WebRTCPeerTransportInit {
+  rtcConfiguration?: RTCConfiguration
+}
+
+export interface WebRTCDirectTransportComponents {
+  peerId: PeerId
+  registrar: Registrar
+  upgrader: Upgrader
+  transportManager: TransportManager
+  peerStore: PeerStore
+}
+
+export class WebRTCDirectTransport implements Transport, Startable {
+  private readonly _started = false
+  private readonly handler?: ConnectionHandler
+
+  constructor (
+    private readonly components: WebRTCDirectTransportComponents,
+    private readonly init: WebRTCPeerTransportInit
+  ) {
+    this._onProtocol = this._onProtocol.bind(this)
+  }
+
+  isStarted () {
+    return this._started
+  }
+
+  async start () {
+    await this.components.registrar.handle(PROTOCOL, (data) => {
+      this._onProtocol(data).catch(err => log.error('failed to handle incoming connect from %p', data.connection.remotePeer, err))
+    })
+  }
+
+  async stop () {
+    await this.components.registrar.unhandle(PROTOCOL)
+  }
+
+  createListener (options: CreateListenerOptions): Listener {
+    return new WebRTCPeerListener(this.components)
+  }
+
+  get [Symbol.toStringTag] (): string {
+    return '@libp2p/webrtc-direct'
+  }
+
+  get [symbol] (): true {
+    return true
+  }
+
+  filter (multiaddrs: Multiaddr[]): Multiaddr[] {
+    return multiaddrs.filter((ma) => {
+      const codes = ma.protoCodes()
+      return codes.includes(CODE)
+    })
+  }
+
+  /*
+   * dial connects to a remote via the circuit relay or any other protocol
+   * and proceeds to upgrade to a webrtc connection.
+   * multiaddr of the form: <multiaddr>/webrtc-direct/p2p/<destination-peer>
+   * For a circuit relay, this will be of the form
+   * <relay address>/p2p/<relay-peer>/p2p-circuit/webrtc-direct/p2p/<destination-peer>
+  */
+  async dial (ma: Multiaddr, options: DialOptions): Promise<Connection> {
+    log.trace('dialing address: ', ma)
+    const addrs = ma.toString().split(TRANSPORT)
+    if (addrs.length !== 2) {
+      // TODO(ckousik): Change to errCode
+      throw new Error('invalid multiaddr')
+    }
+    // look for remote peerId
+    const remoteAddr = multiaddr(addrs[0])
+    const destination = multiaddr(addrs[1])
+
+    const destinationIdString = destination.getPeerId()
+    if (destinationIdString == null) {
+      // TODO(ckousik): Change to errCode
+      throw new Error('bad destination')
+    }
+
+    const controller = new AbortController()
+    if (options.signal == null) {
+      options.signal = controller.signal
+    }
+
+    const connection = await this.components.transportManager.dial(remoteAddr)
+
+    const rawStream = await connection.newStream([PROTOCOL], options)
+
+    const [pc, muxerFactory] = await connect({
+      stream: rawStream,
+      rtcConfiguration: this.init.rtcConfiguration,
+      signal: options.signal
+    })
+
+    rawStream.close()
```

Comment:
Do we need to close this? If the above errors, who closes this?

Reply:
Changed so that either side of the stream will reset if there is an error in processing the SDP. Also, stream will be closed after a connection is successfully upgraded.